File: C++/merge-sort-recursive.cpp
template<typename T>
void merge(T*& arr, int& first, int& mid, int& last)
{
    T temp[size]{0};

    int first1 = first;
    int last1 = mid;
    int first2 = mid + 1;
    int last2 = last;

    int index = first1;

    for (; (first1 <= last1) && (first2 <= last2); ++index)
    {
        if (arr[first1] < arr[first2])
        {
            temp[index] = arr[first1];
            ++first1;
        }
        else
        {
            temp[index] = arr[first2];
            ++first2;
        }
    }

    for (; first1 <= last1; ++first1, ++index)
    {
        temp[index] = arr[first1];
    }

    for (; first2 <= last2; ++first2, ++index)
    {
        temp[index] = arr[first2];
    }

    for (index = first; index <= last; ++index)
    {
        arr[index] = temp[index];
    }

}
template<typename T>
void mergeSort(T*& arr, int first, int last)
{
    if (first < last)
    {
        int mid = (first + last) / 2;
        mergeSort(arr, first, mid);
        mergeSort(arr, mid + 1, last);
        merge(arr, first, mid, last);
    }
}

File: Golang/merge-sort-recursive.go
package main


func merge(arr []int, left int, middle int, right int) {
    n1 := middle - left + 1
    n2 := right - middle

    leftArr := make([]int, n1)
    rightArr := make([]int, n2)

    for i := 0; i < n1; i++ {
        leftArr[i] = arr[left+i]
    }
    for j := 0; j < n2; j++ {
        rightArr[j] = arr[middle+1+j]
    }

    i := 0
    j := 0
    k := left

    for i < n1 && j < n2 {
        if leftArr[i] <= rightArr[j] {
            arr[k] = leftArr[i]
            i++
        } else {
            arr[k] = rightArr[j]
            j++
        }
        k++
    }

    for i < n1 {
        arr[k] = leftArr[i]
        i++
        k++
    }

    for j < n2 {
        arr[k] = rightArr[j]
        j++
        k++
    }
}

func mergeSort(arr []int, left int, right int) {
    if left < right {
        middle := left + (right-left)/2

        mergeSort(arr, left, middle)
        mergeSort(arr, middle+1, right)

        merge(arr, left, middle, right)
    }
}